src/control/server/config/server.go

```diff
@@ -502,37 +509,112 @@ func (cfg *Server) SetNrHugepages(log logging.Logger, mi *common.MemInfo) error
 		log.Debug(msg)
 	}
 
-	if cfgTargetCount <= 0 {
-		return nil // no nvme, no hugepages required
-	}
+	return
+}
 
-	if cfg.DisableHugepages {
-		return FaultConfigHugepagesDisabledWithBdevs
+func (cfg *Server) getMinMaxNrHugepages(log logging.Logger, hpSizeKiB int) (int, int, error) {
+	cfgTargetCount, sysXSCount := cfg.GetTgtCounts(log)
+
+	if cfgTargetCount == 0 {
+		return 0, 0, nil
 	}
 
 	// Calculate minimum number of hugepages for all configured engines.
-	minHugepages, err := storage.CalcMinHugepages(mi.HugepageSizeKiB, cfgTargetCount+sysXSCount)
+	minHugepages, err := storage.CalcMinHugepages(hpSizeKiB, cfgTargetCount+sysXSCount)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	maxTgtCount := largeTargetCount
+	if sysXSCount > 0 {
+		maxTgtCount += largeSysXSCount
+	}
+	maxHugepages, err := storage.CalcMinHugepages(hpSizeKiB, maxTgtCount)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	var msgSysXS string
+	if sysXSCount > 0 {
+		msgSysXS = fmt.Sprintf(" and %d/%d sys-xstreams", sysXSCount, largeSysXSCount)
+	}
+	log.Tracef("calculated min/max %d/%d nr_hugepages based on %d/%d targets%s",
+		minHugepages, maxHugepages, cfgTargetCount, largeTargetCount, msgSysXS)
+
+	if minHugepages > maxHugepages {
+		log.Debugf("config hugepage requirements exceed normal maximum")
+		maxHugepages = minHugepages
+	}
+
+	return minHugepages, maxHugepages, nil
+}
+
+// SetNrHugepages calculates minimum based on total target count if using nvme. If cfg.NrHugepages
+// is set to zero, no hugepage allocation requests will be made to the kernel during service
+// start-up in prepBdevStorage(). Only set non-zero here if a change is required.
+func (cfg *Server) SetNrHugepages(log logging.Logger, mi *common.MemInfo) error {
+	minHugepages, maxHugepages, err := cfg.getMinMaxNrHugepages(log, mi.HugepageSizeKiB)
 	if err != nil {
 		return err
 	}
 
-	// If the config doesn't specify hugepages, use the minimum. Otherwise, validate
-	// that the configured amount is sufficient.
+	// Allow emulated NVMe configurations either with or without hugepages enabled.
+
+	if cfg.DisableHugepages {
+		if cfg.NrHugepages != 0 {
+			return FaultConfigHugepagesDisabledWithNrSet
+		}
+		if cfg.GetBdevConfigs().HaveRealNVMe() {
+			return FaultConfigHugepagesDisabledWithBdevs
+		}
+		if minHugepages != 0 {
+			log.Noticef("hugepages disabled but targets will be assigned to bdevs, " +
+				"this is an atypical situation and caution is advised")
+		}
+
+		// Hugepages disabled and so zero requested in config.
+		return nil
+	} else if minHugepages == 0 {
+		// Enable minimum needed for scanning NVMe on host in discovery mode.
+		if cfg.NrHugepages < scanMinHugepageCount && mi.HugepagesTotal < scanMinHugepageCount {
```

Comment:
I am probably missing something but I was thinking that `mi.HugepagesTotal` was the number of available Huge Pages, and thus `cfg.NrHugePages` could not be greater than this first value.

Reply:
`mi.HugepagesTotal` refers to the number of hugepages already allocated on the system. `cfg.NrHugepages` refers to the configured and required number. This line (L579) is checking whether the configured number needs to be increased to account for a minimum value that will allow NVMe devices to be discovered in the case where the service started with no bdevs  listed in the server config file. If the number of system (already) allocated hugepages is already enough to satisfy this minimum then we don't need to set the configured value as no further allocation is necessary.

Reply:
Thanks for the clarification :-)